flake.nix

```diff
@@ -101,179 +111,135 @@
-                # - when run as a shell hook, we want to be fast, so we let Cargo do incremental compilation, and download dependencies as needed.
-                #   This is mainly used by developers, who care about speed at the expense of hermeticity.
-                # - when run as a flake check, we want to be fully reproducible, so we Nixify Cargo dependency resolution.
-                #   This is mainly used by CI, where the extra time is not such a big deal.
-                offlineOpts = pkgs.lib.optionalString isHermetic "--frozen --offline";
-
-                # When running in hermetic mode, all Rust dependencies must be already present.
-                # Hence we can't just call `cargo` directly, as it would try to download dependencies.
-                # Hence we do the following:
-                # - `source ${nickel.inputDerivation}` to set the environment variables `buildInputs` and `stdenv` just like in a Nickel build.
-                #   The `buildInputs` will contain the path to Nixified Rust dependencies (a.k.a. `cargoHome`), as well as other build inputs, like `Security` for Darwin.
-                #   ⚠️ This will also completely unset `PATH`, and *update* various environment variables used in downstream scripts, like `TMPDIR`.
-                # - `source $stdenv/setup` does a lot of Nix magic, including loading in the `PATH` everything in the `buildInputs` environment variable.
-                #
-                # 💡 The other commands are needed for Darwin builds to undo some changes of `source ${nickel.inputDerivation}`.
-                #    For Linux, these extra commands are unnecessary (because their value is "updated" from `/build` to `/build`), though harmless.
-                nickelBuildInputsSetup = pkgs.lib.optionalString isHermetic ''
-                  nix_build_top=$NIX_BUILD_TOP
-                  source ${nickel.inputDerivation}
-                  NIX_BUILD_TOP=$nix_build_top
-                  TEMP=$nix_build_top
-                  TEMPDIR=$nix_build_top
-                  TMP=$nix_build_top
-                  TMPDIR=$nix_build_top
-                  source $stdenv/setup
-                '';
-              in
-              {
-                enable = true;
-                files = "\\.rs$";
-                pass_filenames = false;
-                entry =
-                  "${pkgs.writeShellScript "clippy-hook" ''
-                    ${nickelBuildInputsSetup}
-                    cargo clippy --workspace ${offlineOpts} -- --no-deps --deny warnings ${allowOpts}
-                  ''}";
-              };
+          # Args passed to all `cargo` invocations by Crane.
+          cargoExtraArgs = "--frozen --offline --workspace";
+
+        in
+        rec {
+          # Build *just* the cargo dependencies, so we can reuse all of that work (e.g. via cachix) when running in CI
+          cargoArtifacts = craneLib.buildDepsOnly {
+            inherit
+              src
+              cargoExtraArgs;
```

Comment:
I like the declarative aspect of Crane: I find it easier to understand what influences what in the chain of dependencies.